sdk/storage/Azure.Storage.Common/src/Shared/AesGcm/AllowUnsafeBlocks/AeadCommon.cs

```diff
@@ -0,0 +1,130 @@
+                    ciphertext.Length,
+                    out int ciphertextBytesWritten,
+                    0);
+
+                Debug.Assert(plaintext.Length == ciphertextBytesWritten);
+
+                if (ntStatus != Interop.BCrypt.NTSTATUS.STATUS_SUCCESS)
+                {
+                    throw Interop.BCrypt.CreateCryptographicException(ntStatus);
+                }
+            }
+        }
+
+        internal static unsafe void Decrypt(
+            Interop.BCrypt.SafeKeyHandle keyHandle,
+            ReadOnlySpan<byte> nonce,
+            ReadOnlySpan<byte> associatedData,
+            ReadOnlySpan<byte> ciphertext,
+            ReadOnlySpan<byte> tag,
+            Span<byte> plaintext,
+            bool clearPlaintextOnFailure)
+        {
+            // bcrypt sometimes misbehaves when given nullptr buffers; ensure non-nullptr
+            fixed (byte* plaintextBytes = &GetNonNullPinnableReference(plaintext))
+            fixed (byte* nonceBytes = &GetNonNullPinnableReference(nonce))
+            fixed (byte* ciphertextBytes = &GetNonNullPinnableReference(ciphertext))
+            fixed (byte* tagBytes = &GetNonNullPinnableReference(tag))
+            fixed (byte* associatedDataBytes = &GetNonNullPinnableReference(associatedData))
+            {
+                var authInfo = Interop.BCrypt.BCRYPT_AUTHENTICATED_CIPHER_MODE_INFO.Create();
+                authInfo.pbNonce = nonceBytes;
+                authInfo.cbNonce = nonce.Length;
+                authInfo.pbTag = tagBytes;
+                authInfo.cbTag = tag.Length;
+                authInfo.pbAuthData = associatedDataBytes;
+                authInfo.cbAuthData = associatedData.Length;
+
+                Interop.BCrypt.NTSTATUS ntStatus = Interop.BCrypt.BCryptDecrypt(
+                    keyHandle,
+                    ciphertextBytes,
+                    ciphertext.Length,
+                    new IntPtr(&authInfo),
+                    null,
+                    0,
+                    plaintextBytes,
+                    plaintext.Length,
+                    out int plaintextBytesWritten,
+                    0);
+
+                Debug.Assert(ciphertext.Length == plaintextBytesWritten);
+
+                switch (ntStatus)
+                {
+                    case Interop.BCrypt.NTSTATUS.STATUS_SUCCESS:
+                        return;
+                    case Interop.BCrypt.NTSTATUS.STATUS_AUTH_TAG_MISMATCH:
+                        if (clearPlaintextOnFailure)
+                        {
+                            plaintext.Clear();
+                            //CryptographicOperations.ZeroMemory(plaintext);
+                        }
+                        Errors.CryptographyAuthTagMismatch();
+                        return;
+                    default:
+                        //throw CreateCryptographicException(ntStatus);
+                        throw new CryptographicException($"Error: 0x{(int)ntStatus:X8}");
```

Comment:
Same here

Reply:
Yeah, both here and Encrypt represent "we got an error code that we don't understand".  For Encrypt, the documented return codes are https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptencrypt#return-value

All of the documented failures were checked by us in preconditions to make for nicer exceptions.  Anything else is a surprise.

What we do in the framework is call into FormatMessage to ask the OS for the text associated with the error code.  It's probably good enough to share just the error code, and if people see them, get confused by them, and report them, then maybe there are some more returns worth handling.